### PR TITLE
Allow `make -C testuite promote` to take `TEST` and `LIST` variables

### PR DESCRIPTION
**Pull request by antalsz**
The status quo is that you can run any of

```
make -C testsuite one DIR=tests/typing-modules
make -C testsuite one TEST=tests/typing-modules/printing.ml
make -C testsuite one LIST=file
```

if you want to run tests, but you have to run

```
make -C testsuite promote DIR=tests/typing-modules
```

if you want to promote new test output.  This PR attempts to update the machinery for `promote` to be the same as the machinery for `one`, so you can now say

```
make -C testsuite promote TEST=tests/typing-modules/printing.ml
make -C testsuite promote LIST=file
```

as well.

This also fixes an issue where `promote` didn't correctly depend on (`lib` and?) `tools`.

---

I believe this code preserves the existing behavior, as `one` was already checking for a `PROMOTE` variable, except that we change the fallback case to no longer try to `make promote` in the given directory; however, this fallback is currently unused, as the only `Makefile` in `testsuite/tests` is `testsuite/tests/memory-model/Makefile`, which doesn't do any promotion.

(It's not clear to me if this is the sort of change that needs a `Changes` entry, but I'm happy to write one if it does)

### CODE REVIEW

**Review by gasche — APPROVED** [review]
This looks nice, thanks! Could you maybe consider adding mentions of LIST, TEST to `testsuite/HACKING.adoc`? (They don't need to be specific to `one` or `promote` given that both now work the same way.)

File: testsuite/Makefile [hunk 1]
```diff
@@ -253,7 +256,7 @@ one: lib tools
 	@case "$(words $(DIR) $(LIST) $(TEST))" in \
    0) echo 'No value set for variable DIR, LIST or TEST'>&2; exit 1;; \
    1) exit 0;; \
-   *) echo 'Please specify just one of DIR, LIST or TEST'>&2; exit 1;; \
+   *) echo 'Please specify exactly one of DIR, LIST or TEST'>&2; exit 1;; \
```
Comment (dra27) [inline testsuite/Makefile:259]: This isn't necessary?

File: testsuite/Makefile [hunk 2]
```diff
@@ -155,7 +155,10 @@ default:
 	@echo "  one TEST=f      launch just this single test"
 	@echo "  one DIR=p       launch the tests located in path p"
 	@echo "  one LIST=f      launch the tests listed in f (one per line)"
-	@echo "  promote DIR=p   promote the reference files for the tests in p"
+	@echo "  promote TEST=f  promote the output for this single test"
+	@echo "  promote DIR=p   promote the output for the tests located in path p"
+	@echo "  promote LIST=f  promote the output for the tests listed in f (one \
+	per line)"
```
Comment (dra27) [inline testsuite/Makefile:161]: This is getting a bit wordy with 6 lines, it could possibly be refactored something like:

```
	@echo "  one DIR=p       launch the tests located in path p"
	@echo "  promote DIR=p   promote the output for the tests located in path p"
	@echo "                  (both one and promote accept TEST=f for a single test and
	@echo "                  LIST=f for a list of tests in f (one per line)"
```

**Review by dra27 — APPROVED** [review]
Possible rewording of the help text just to reduce the number of lines, but I likewise it's a good change!

File: testsuite/Makefile (hunk 1 above)
Comment (antalsz) [reply testsuite/Makefile:259]: To me, the former wording implied you could leave out *all* of `DIR`, `LIST`, and `TEST`, which isn't the case

File: testsuite/Makefile (hunk 2 above)
Comment (antalsz) [reply testsuite/Makefile:161]: What about something like this?
```
Available targets:
  all                         launch all tests
  all-foo                     launch all tests beginning with foo
  parallel                    launch all tests using GNU parallel
  parallel-foo                launch all tests beginning with foo using GNU parallel
  one (TEST|DIR|LIST)=x       launch the specified tests
  promote (TEST|DIR|LIST)=x   promote the output for the specified tests
    ... TEST=f                specify this single test
    ... DIR=p                 specify the tests located in path p
    ... LIST=f                specify the tests listed in f (one per line)
  lib                         build library modules
  tools                       build test tools
  clean                       delete generated files
  report                      print the report for the last execution
```
This has the "parallel-foo" line at over 80 characters though, and it only saves one line.  Even the version you propose only saves two lines, and it hides two very useful options inside a parenthetical.  I'm not sure there's a good option that's clearer than just listing all the targets directly.  We could also try grouping them into sections, or adding line breaks to allow for more easily following what's there?

**Comment by antalsz** [comment]
@gasche I added documentation to `testsuite/HACKING.adoc`.  Also, does this change require updating `Changes.md` or not?

File: testsuite/Makefile (hunk 2 above)
Comment (dra27) [reply testsuite/Makefile:161]: I like your proposal! By "wordy" I was meaning the repetition of "one", "one", "one"; "promote", "promote", "promote" which you avoid

File: testsuite/Makefile (hunk 1 above)
Comment (dra27) [reply testsuite/Makefile:259]: That has an explicit error two lines earlier, though!

**Comment by dra27** [comment]
The changes to `testsuite/HACKING.adoc` look great, thank you! It's entirely up to you about `Changes` - testsuite changes can get put in the "Internal/compiler-libs changes" section. On the other hand, if you're not bothered to have an entry, one of us can add the label which suppresses the hygiene job.

File: testsuite/Makefile (hunk 2 above)
Comment (antalsz) [reply testsuite/Makefile:161]: The precise text I went with is

```
Available targets:
  all                        launch all tests
  all-foo                    launch all tests beginning with foo
  parallel                   launch all tests using GNU parallel
  parallel-foo               launch all tests beginning with foo using GNU parallel
  one (TEST|DIR|LIST)=x      launch the specified tests ...
  promote (TEST|DIR|LIST)=x  promote the output for the specified tests ...
    ... TEST=f                 ... the single test in the file f
    ... DIR=d                  ... the tests located in the directory d
    ... LIST=f                 ... the tests listed in the file f (one per line)
  lib                        build library modules
  tools                      build test tools
  clean                      delete generated files
  report                     print the report for the last execution

all*, parallel* and list can automatically re-run failed test
directories if MAX_TESTSUITE_DIR_RETRIES permits
(default value = )
```

My only potential concern is that the `parallel-foo` line is >80 columns (by 3); my only proposed fix is changing "launch" to "run" as the verb for what we do with tests, but I'm open to other ideas (e.g., `starting` is 1 character less than `beginning`).

File: testsuite/Makefile (hunk 1 above)
Comment (antalsz) [reply testsuite/Makefile:259]: It does, but I think any error message you get should be clear in isolation – why make somebody round-trip through two errors instead of one?  It's also possible that the `0` and `*` cases could share an error message, too...

**Comment by gasche** [comment]
I think we spent an admirable bit of effort on this patch; thank you both :-)
